PythonVisualizations/Heap.py

```diff
@@ -181,13 +184,212 @@ def siftUp(self, i):
+
+        self.cleanUp(callEnviron)
+              
     # lets user input an int argument that determines max size of the Heap
     def newArray(self):
         #gets rid of old elements in the list
         del self.list[:]
         self.size = 2
         self.display()
+    
+
+    def peek(self):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
 
+        # draw output box
+        canvasDimensions = self.widgetDimensions(self.canvas)
+        spacing = self.CELL_WIDTH * 3 // 4
+        padding = 10
+        boxSize = 1
+        peekIndex = 0
+        # changed the values that were added to the variables to get the output box in a desirable location
+        outputBox = self.canvas.create_rectangle(
+            (self.HEAP_X0 * 2 + padding),
+            canvasDimensions[1] - self.CELL_WIDTH - padding,
+            (self.HEAP_X0 * 2 + boxSize * spacing + padding * 2),
+            canvasDimensions[1] - padding,
+            fill=self.OPERATIONS_BG)
+        callEnviron.add(outputBox)
+
+        outputBoxCoords = self.canvas.coords(outputBox)
+        midOutputBox = (outputBoxCoords[3] + outputBoxCoords[1]) // 2
+        # create the value to move to output box
+        valueOutput = self.copyCanvasItem(self.list[peekIndex].display_val)
+        valueList = (valueOutput,)
+        callEnviron.add(valueOutput)
+        # move value to output box
+        # self.highlightCodeTags('print', callEnviron)
+        toPositions = (outputBoxCoords[0] + padding / 2 + (peekIndex + 1 / 2) * spacing,
+                        midOutputBox)
+        self.moveItemsTo(valueList, (toPositions,), sleepTime=.02)
+        # make the value 25% smaller
+        newFont = (self.VALUE_FONT[0], int(self.VALUE_FONT[1] * .75))
+        self.canvas.itemconfig(valueOutput, font=newFont)
+
+        # add label to output box
+        labelX = self.HEAP_X0 * 2 + boxSize * spacing + padding * 5
+        labelY = canvasDimensions[1] - padding * 2
+
+        boxLabel = self.canvas.create_text(
+            labelX, labelY, text="root", font=self.VARIABLE_FONT)
+        callEnviron.add(boxLabel)
+        self.highlightCodeTags([], callEnviron)
+        # finish animation
+        self.cleanUp(callEnviron)
+    
+    def removeMax(self):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+
+        # remove first element from list
+        n = self.list[0]
+        self.list[0] = None
+        self.size -= 1
+        # if the array is now empty, delete the NoneType place holder from the array
+        if self.size < 1: del self.list[:]
+        # Slide value rectangle up and off screen
+        items = (n.display_shape, n.display_val)
+        toPositions = (100, -100, 100, -50)
+        callEnviron |= set(items)
+        self.moveItemsBy(items, delta=(0, -self.canvas.coords(n.display_shape)[3]), steps=self.CELL_HEIGHT,
+                        sleepTime=.05)
+        self.canvas.delete(items)
```

Comment:
I don't think the `toPositions` variable defined in line 338 is doing anything.

In Issue #77, it asks for:
>  The Remove Max should also animate copying the first
> element of the array to the Root result, and then animate the process
> of moving the last element to the first element of the array, reducing
> `nItems`, and then sifting the first item down.

So instead of moving the root value off canvas it should move the value to the same place that the Peek operation does.  It should not delete the value items as it is doing in line 342.

Reply:
So instead of lines 337-342, I decided to try calling the peek() method- before removing the first cell from the implicit memory of the list- because I didn't want to duplicate all the code from peek() in this method as well. The problem with that was that the cleanup() method at the end of peek() makes it immediately disappear. Is there a way to get around that? I was initially thinking of taking most of the peek() code out of peek() and turning it into a more broad method, displayOnBottom() or something like that, and not include the call environment in that method, so that when peek() or removeFromMax() calls that method, they can decide if they want a call environment, but I don't think the callEnvironment works like that so it wouldn't actually work. Any thoughts on the best way to approach this?

Reply:
That's a good instinct to re-use the code of `peek()` for the animation of the removal of the root of the heap.  To see what's best in terms of code sharing, we should look at the differences between the two use cases:

**Peek**
  * Creates the output area rectangle and its label and stores them in a call environment
  * Animates the movement of the text item for the value being output
  * Leaves the output value in the call environment
  * Cleans up its call environment at the end
  * In the future, it will show the code for the `peek()` operation with highlights

**Remove Max (before moving the last item to the root)**
  * Usually animates the movement of text items with their colored background to show assignment.  Assuming we do animate the colored background, we would need to make it disappear once the value is in the output area.
  * Needs to put the output area rectangle, its label, and the removed max value in its own call environment for later clean up

I would suggest making a common routine that both `peek()` and  `removeMax()` call to do the common work.  It would have a `callEnviron` parameter in which it can put the new canvas items it creates.  It would need another parameter, `moveShape` or something, to be a flag indicating whether to move the background shape along with the text item to the output rectangle.  The background shape should be put at a lower stacking level on the canvas than the output rectangle so that it disappears behind the rectangle while the text stays visible on top (see the canvas `tag_lower()` method if that's unfamiliar).

The common routine would not call `cleanUp()`.  It leaves that to the method that called it.  The `peek()` method would display the code for its operation, highlight the statement that gets the max value, and then call the common routine with the `moveShape` flag as false to do the animation.  The 'removeMax()` method would display its own code, highlight the assignment statement that gets the max value, and then call the common routine with the `moveShape` flag as true.  It would leave the items created by the common routine on the canvas while it animates the rest of the removal operation.

That's one suggestion.  There might be better ways of doing it.  When we incorporate the tree visualization of the heap, there could be other issues (like animating the movement from both the tree and the array).